cli/src/commands/undo.rs

```diff
@@ -72,56 +60,69 @@ pub fn cmd_undo(ui: &mut Ui, command: &CommandHelper, args: &UndoArgs) -> Result
+    // - Repeat the process of following undo-operations to the operations they
+    //   undid until the first undoable operation is found - then undo it.
     //
-    //    C (undo of B)
-    // @  B (`bad_op` = undo of A)
-    // ○  A
+    // This described behavior leads to "jumping over" old undo-stacks if the
+    // current one grows into it. For example, Consider the this op-log example:
     //
-    // An exception is made for when the user specified the immediately preceding
-    // `undo` with an op set. In this situation, the user's intent is clear, so
-    // a warning is not shown.
+    // * F "undo A" ---+
+    // |               |
+    // * E "undo D" -+ |
+    // |             | |
+    // * D   <-------+ |
+    // |               |
+    // * C "undo B" -+ |
+    // |             | |
+    // * B   <-------+ |
+    // |               |
+    // * A   <---------+
     //
-    // Note that undoing an older `undo` does not constitute a "double undo". For
-    // example, the current `undo` (D) here reverts an `undo` B that is not the
-    // immediately preceding operation (C). A warning is not shown in this case.
+    // It was produced by the following sequence of events:
+    // - do normal operation A
+    // - do normal operation B
+    // - undo B
+    // - do normal operation D
+    // - undo D
+    // - undo A
     //
-    //    D (undo of B)
-    // @  C (unrelated operation)
-    // ○  B (`bad_op` = undo of A)
-    // ○  A
-    if args.operation == "@"
-        && resets_view_of(&bad_op, &parent_of_bad_op)?
-        && bad_op.metadata().description == tx_description(&parent_of_bad_op)
+    // Notice that running `undo` after having undone D leads to A being undone
+    // (as opposed to C). The undo-stack spanning B and C was "jumped over".
+    //
+    while let Some(id_of_undone_op) = op_to_undo
+        .metadata()
+        .description
+        .strip_prefix(UNDO_OP_DESC_PREFIX)
     {
-        writeln!(
-            ui.warning_default(),
-            "The second-last `jj undo` was reverted by the latest `jj undo`. The repo is now in \
-             the same state as it was before the second-last `jj undo`."
-        )?;
-        writeln!(
-            ui.hint_default(),
-            "To undo multiple operations, use `jj op log` to see past states and `jj op restore` \
-             to restore one of these states."
-        )?;
+        let undone_op = workspace_command.resolve_single_op(id_of_undone_op)?;
+        op_to_undo = match undone_op.parents().at_most_one() {
+            Ok(Some(parent_of_undone_op)) => parent_of_undone_op?,
+            Ok(None) => return Err(user_error("Cannot undo root operation")),
+            Err(_) => return Err(user_error("Cannot undo a merge operation")),
+        };
     }
 
+    let args = OperationRevertArgs {
+        operation: op_to_undo.id().to_string(),
+        what: args.what.clone(),
+    };
+    cmd_op_revert_with_tx_description(ui, command, &args, tx_description)?;
```

Comment:
nit: I think we can use the "op restore" logic here to clarify that we aren't creating new state by merging reverse operation.